File: packages/sdk-python/src/agentops/correlation/manager.py
"""Trace manager for multi-agent correlation."""

import random
import time
import uuid
from typing import Any
from urllib.parse import quote, unquote

from .types import (
    AgentInfo,
    CorrelationConfig,
    SpanInfo,
    TraceContext,
    TraceStats,
)


def generate_trace_id() -> str:
    """Generate a unique trace ID."""
    return f"tr_{uuid.uuid4().hex[:28]}"


def generate_span_id() -> str:
    """Generate a unique span ID."""
    return f"sp_{uuid.uuid4().hex[:16]}"


class TraceManager:
    """Manages distributed traces across multi-agent systems."""

    def __init__(self, config: CorrelationConfig | None = None):
        self.config = config or CorrelationConfig()
        self._active_spans: dict[str, SpanInfo] = {}
        self._completed_spans: dict[str, list[SpanInfo]] = {}
        self._current_context: TraceContext | None = None

        if self.config.agent is None:
            self.config.agent = AgentInfo(
                agent_id=f"agent_{int(time.time())}",
                name="default-agent",
            )

    @property
    def is_enabled(self) -> bool:
        return self.config.enabled

    @property
    def agent(self) -> AgentInfo:
        return self.config.agent  # type: ignore

    def start_trace(
        self,
        name: str,
        attributes: dict[str, Any] | None = None,
    ) -> SpanInfo:
        """Start a new trace (root span)."""
        context = self._create_root_context()
        self._current_context = context
        return self._start_span_with_context(context, name, attributes)

    def continue_trace(
        self,
        incoming_context: TraceContext,
        name: str,
        attributes: dict[str, Any] | None = None,
    ) -> SpanInfo:
        """Continue a trace from an incoming context."""
        child_context = self._create_child_context(incoming_context)
        self._current_context = child_context
        return self._start_span_with_context(child_context, name, attributes)

    def start_span(
        self,
        name: str,
        attributes: dict[str, Any] | None = None,
    ) -> SpanInfo:
        """Start a new span within the current trace."""
        if not self._current_context:
            return self.start_trace(name, attributes)

        child_context = self._create_child_context(self._current_context)
        return self._start_span_with_context(child_context, name, attributes)

    def end_span(
        self,
        span_id: str,
        status: str = "ok",
        error_message: str | None = None,
    ) -> SpanInfo | None:
        """End a span."""
        span = self._active_spans.get(span_id)
        if not span:
            return None

        span.end_time = int(time.time() * 1000)
        span.duration_ms = span.end_time - span.start_time
        span.status = status
        if error_message:
            span.error_message = error_message

        del self._active_spans[span_id]

        if span.trace_id not in self._completed_spans:
            self._completed_spans[span.trace_id] = []
        self._completed_spans[span.trace_id].append(span)

        return span

    def get_current_context(self) -> TraceContext | None:
        return self._current_context

    def set_current_context(self, context: TraceContext) -> None:
        self._current_context = context

    def inject_context(
        self,
        headers: dict[str, str] | None = None,
    ) -> dict[str, str]:
        """Inject trace context into headers for outgoing requests."""
        result = dict(headers or {})
        
        if not self._current_context:
            return result

        ctx = self._current_context
        hdrs = self.config.propagation_headers

        result[hdrs["trace_id"]] = ctx.trace_id
        result[hdrs["span_id"]] = ctx.span_id
        result[hdrs["sampled"]] = "1" if ctx.sampled else "0"

        if ctx.parent_span_id:
            result[hdrs["parent_span_id"]] = ctx.parent_span_id

        if self.config.propagate_baggage and ctx.baggage:
            baggage_entries = [
                f"{quote(k)}={quote(v)}"
                for k, v in list(ctx.baggage.items())[:self.config.max_baggage_items]
            ]
            if baggage_entries:
                result[hdrs["baggage"]] = ",".join(baggage_entries)

        return result

    def extract_context(
        self,
        headers: dict[str, str | None],
    ) -> TraceContext | None:
        """Extract trace context from incoming headers."""
        hdrs = self.config.propagation_headers

        def get_header(name: str) -> str | None:
            return headers.get(name) or headers.get(name.lower())

        trace_id = get_header(hdrs["trace_id"])
        if not trace_id:
            return None

        span_id = get_header(hdrs["span_id"]) or generate_span_id()
        parent_span_id = get_header(hdrs["parent_span_id"])
        sampled = get_header(hdrs["sampled"]) != "0"
        baggage = self._parse_baggage(get_header(hdrs["baggage"]))

        return TraceContext(
            trace_id=trace_id,
            span_id=span_id,
            parent_span_id=parent_span_id,
            sampled=sampled,
            baggage=baggage,
        )

    def add_baggage(self, key: str, value: str) -> None:
        """Add a baggage item to the current context."""
        if not self._current_context:
            return

        if len(self._current_context.baggage) >= self.config.max_baggage_items:
            return

        self._current_context.baggage[key] = value

    def get_baggage(self, key: str) -> str | None:
        """Get a baggage item from the current context."""
        if not self._current_context:
            return None
        return self._current_context.baggage.get(key)

    def get_trace_stats(self, trace_id: str) -> TraceStats | None:
        """Get statistics for a trace."""
        spans = self._completed_spans.get(trace_id)
        if not spans:
            return None

        agent_stats: dict[str, dict[str, int]] = {}
        agents: set[str] = set()
        error_count = 0
        total_duration_ms = 0

        for span in spans:
            agents.add(span.agent_id)

            if span.agent_id not in agent_stats:
                agent_stats[span.agent_id] = {
                    "span_count": 0,
                    "total_duration_ms": 0,
                    "error_count": 0,
                }

            agent_stats[span.agent_id]["span_count"] += 1
            agent_stats[span.agent_id]["total_duration_ms"] += span.duration_ms or 0

            if span.status == "error":
                agent_stats[span.agent_id]["error_count"] += 1
                error_count += 1

            end = span.end_time or 0
            total_duration_ms = max(total_duration_ms, end - spans[0].start_time)

        return TraceStats(
            span_count=len(spans),
            agent_count=len(agents),
            total_duration_ms=total_duration_ms,
            critical_path_ms=self._calculate_critical_path(spans),
            error_count=error_count,
            agent_stats=agent_stats,
        )

    def get_trace_spans(self, trace_id: str) -> list[SpanInfo]:
        return self._completed_spans.get(trace_id, [])

    def clear_trace(self, trace_id: str) -> None:
        self._completed_spans.pop(trace_id, None)

    def _create_root_context(self) -> TraceContext:
        sampled = random.random() < self.config.sampling_rate
        return TraceContext(
            trace_id=generate_trace_id(),
            span_id=generate_span_id(),
            sampled=sampled,
            baggage={},
        )

    def _create_child_context(self, parent: TraceContext) -> TraceContext:
        return TraceContext(
            trace_id=parent.trace_id,
            span_id=generate_span_id(),
            parent_span_id=parent.span_id,
            sampled=parent.sampled,
            baggage=dict(parent.baggage),
        )

    def _start_span_with_context(
        self,
        context: TraceContext,
        name: str,
        attributes: dict[str, Any] | None,
    ) -> SpanInfo:
        span = SpanInfo(
            span_id=context.span_id,
            parent_span_id=context.parent_span_id,
            trace_id=context.trace_id,
            name=name,
            agent_id=self.agent.agent_id,
            start_time=int(time.time() * 1000),
            status="in_progress",
            attributes=attributes or {},
        )

        self._active_spans[span.span_id] = span
        self._current_context = context

        return span

    def _parse_baggage(self, baggage_str: str | None) -> dict[str, str]:
        if not baggage_str:
            return {}

        result: dict[str, str] = {}
        for entry in baggage_str.split(",")[:self.config.max_baggage_items]:
            parts = entry.split("=", 1)
            if len(parts) == 2:
                try:
                    result[unquote(parts[0].strip())] = unquote(parts[1].strip())
                except Exception:
                    pass
        return result

    def _calculate_critical_path(self, spans: list[SpanInfo]) -> int:
        if not spans:
            return 0

        span_map = {s.span_id: s for s in spans}
        children: dict[str, list[SpanInfo]] = {}
        root_span: SpanInfo | None = None

        for span in spans:
            if not span.parent_span_id:
                root_span = span
            else:
                if span.parent_span_id not in children:
                    children[span.parent_span_id] = []
                children[span.parent_span_id].append(span)

        if not root_span:
            return max(s.duration_ms or 0 for s in spans)

        def find_longest_path(span: SpanInfo) -> int:
            child_spans = children.get(span.span_id, [])
            if not child_spans:
                return span.duration_ms or 0

            max_child = max(find_longest_path(c) for c in child_spans)
            return (span.duration_ms or 0) + max_child

        return find_longest_path(root_span)

File: packages/sdk-ts/src/correlation/manager.ts
/**
 * AgentOps SDK - Trace Manager
 *
 * Manages distributed traces across multi-agent systems.
 */

import type {
  TraceContext,
  SpanInfo,
  AgentInfo,
  CorrelationConfig,
  ResolvedCorrelationConfig,
  TraceStats,
} from "./types.js";
import { ContextPropagator } from "./propagator.js";
import { now, generateEventId } from "../utils.js";

const DEFAULT_CORRELATION_CONFIG: ResolvedCorrelationConfig = {
  enabled: false,
  agent: {
    agentId: `agent_${Date.now()}`,
    name: "default-agent",
  },
  samplingRate: 1.0,
  propagateBaggage: true,
  maxBaggageItems: 64,
  propagationHeaders: {
    traceId: "x-agentops-trace-id",
    spanId: "x-agentops-span-id",
    parentSpanId: "x-agentops-parent-span-id",
    sampled: "x-agentops-sampled",
    baggage: "x-agentops-baggage",
  },
};

export class TraceManager {
  private readonly config: ResolvedCorrelationConfig;
  private readonly propagator: ContextPropagator;
  private activeSpans: Map<string, SpanInfo> = new Map();
  private completedSpans: Map<string, SpanInfo[]> = new Map();
  private currentContext: TraceContext | null = null;

  constructor(config?: CorrelationConfig) {
    this.config = {
      ...DEFAULT_CORRELATION_CONFIG,
      ...config,
      agent: config?.agent ?? DEFAULT_CORRELATION_CONFIG.agent,
      propagationHeaders: {
        ...DEFAULT_CORRELATION_CONFIG.propagationHeaders,
        ...config?.propagationHeaders,
      },
    };
    this.propagator = new ContextPropagator(this.config);
  }

  get isEnabled(): boolean {
    return this.config.enabled;
  }

  get agent(): AgentInfo {
    return this.config.agent;
  }

  /**
   * Start a new trace (root span)
   */
  startTrace(name: string, attributes?: Record<string, unknown>): SpanInfo {
    const context = this.propagator.createRootContext();
    this.currentContext = context;

    return this.startSpanWithContext(context, name, attributes);
  }

  /**
   * Continue a trace from an incoming context
   */
  continueTrace(
    incomingContext: TraceContext,
    name: string,
    attributes?: Record<string, unknown>,
  ): SpanInfo {
    const childContext = this.propagator.createChildContext(incomingContext);
    this.currentContext = childContext;

    return this.startSpanWithContext(childContext, name, attributes);
  }

  /**
   * Start a new span within the current trace
   */
  startSpan(name: string, attributes?: Record<string, unknown>): SpanInfo {
    if (!this.currentContext) {
      return this.startTrace(name, attributes);
    }

    const childContext = this.propagator.createChildContext(
      this.currentContext,
    );
    return this.startSpanWithContext(childContext, name, attributes);
  }

  /**
   * End a span
   */
  endSpan(
    spanId: string,
    status: "ok" | "error" = "ok",
    errorMessage?: string,
  ): SpanInfo | null {
    const span = this.activeSpans.get(spanId);
    if (!span) {
      return null;
    }

    span.endTime = now();
    span.durationMs = span.endTime - span.startTime;
    span.status = status;
    if (errorMessage) {
      span.errorMessage = errorMessage;
    }

    this.activeSpans.delete(spanId);

    // Store in completed spans by trace
    const traceSpans = this.completedSpans.get(span.traceId) || [];
    traceSpans.push(span);
    this.completedSpans.set(span.traceId, traceSpans);

    return span;
  }

  /**
   * Get current trace context for propagation
   */
  getCurrentContext(): TraceContext | null {
    return this.currentContext;
  }

  /**
   * Set current context (e.g., from extracted headers)
   */
  setCurrentContext(context: TraceContext): void {
    this.currentContext = context;
  }

  /**
   * Inject trace context into headers for outgoing requests
   */
  injectContext(headers: Record<string, string> = {}): Record<string, string> {
    if (!this.currentContext) {
      return headers;
    }
    return this.propagator.inject(this.currentContext, headers);
  }

  /**
   * Extract trace context from incoming headers
   */
  extractContext(
    headers: Record<string, string | undefined>,
  ): TraceContext | null {
    return this.propagator.extract(headers);
  }

  /**
   * Add baggage to current context
   */
  addBaggage(key: string, value: string): void {
    if (this.currentContext) {
      this.currentContext = this.propagator.addBaggage(
        this.currentContext,
        key,
        value,
      );
    }
  }

  /**
   * Get baggage from current context
   */
  getBaggage(key: string): string | undefined {
    if (!this.currentContext) {
      return undefined;
    }
    return this.propagator.getBaggage(this.currentContext, key);
  }

  /**
   * Record an agent-to-agent call
   */
  recordAgentCall(
    _targetAgentId: string,
    _callType: "sync" | "async" = "sync",
  ): {
    eventId: string;
    headers: Record<string, string>;
  } {
    const eventId = generateEventId();
    const headers = this.injectContext();

    return { eventId, headers };
  }

  /**
   * Get statistics for a trace
   */
  getTraceStats(traceId: string): TraceStats | null {
    const spans = this.completedSpans.get(traceId);
    if (!spans || spans.length === 0) {
      return null;
    }

    const agentStats: TraceStats["agentStats"] = {};
    let totalDurationMs = 0;
    let errorCount = 0;
    const agents = new Set<string>();

    for (const span of spans) {
      agents.add(span.agentId);

      if (!agentStats[span.agentId]) {
        agentStats[span.agentId] = {
          spanCount: 0,
          totalDurationMs: 0,
          errorCount: 0,
        };
      }

      agentStats[span.agentId].spanCount++;
      agentStats[span.agentId].totalDurationMs += span.durationMs || 0;

      if (span.status === "error") {
        agentStats[span.agentId].errorCount++;
        errorCount++;
      }

      totalDurationMs = Math.max(
        totalDurationMs,
        (span.endTime || 0) - spans[0].startTime,
      );
    }

    // Calculate critical path (simplified: longest path from root to leaf)
    const criticalPathMs = this.calculateCriticalPath(spans);

    return {
      spanCount: spans.length,
      agentCount: agents.size,
      totalDurationMs,
      criticalPathMs,
      errorCount,
      agentStats,
    };
  }

  /**
   * Get all spans for a trace
   */
  getTraceSpans(traceId: string): SpanInfo[] {
    return this.completedSpans.get(traceId) || [];
  }

  /**
   * Clear completed spans for a trace
   */
  clearTrace(traceId: string): void {
    this.completedSpans.delete(traceId);
  }

  private startSpanWithContext(
    context: TraceContext,
    name: string,
    attributes?: Record<string, unknown>,
  ): SpanInfo {
    const span: SpanInfo = {
      spanId: context.spanId,
      parentSpanId: context.parentSpanId,
      traceId: context.traceId,
      name,
      agentId: this.config.agent.agentId,
      startTime: now(),
      status: "in_progress",
      attributes,
    };

    this.activeSpans.set(span.spanId, span);
    this.currentContext = context;

    return span;
  }

  private calculateCriticalPath(spans: SpanInfo[]): number {
    if (spans.length === 0) return 0;

    // Build span tree
    const spanMap = new Map<string, SpanInfo>();
    const children = new Map<string, SpanInfo[]>();
    let rootSpan: SpanInfo | null = null;

    for (const span of spans) {
      spanMap.set(span.spanId, span);
      if (!span.parentSpanId) {
        rootSpan = span;
      } else {
        const siblings = children.get(span.parentSpanId) || [];
        siblings.push(span);
        children.set(span.parentSpanId, siblings);
      }
    }

    if (!rootSpan) {
      return spans.reduce((max, s) => Math.max(max, s.durationMs || 0), 0);
    }

    // DFS to find longest path
    const findLongestPath = (span: SpanInfo): number => {
      const childSpans = children.get(span.spanId) || [];
      if (childSpans.length === 0) {
        return span.durationMs || 0;
      }

      const maxChildPath = Math.max(
        ...childSpans.map((c) => findLongestPath(c)),
      );
      return (span.durationMs || 0) + maxChildPath;
    };

    return findLongestPath(rootSpan);
  }
}
